File: src/degenbot/matching/internal_matcher.py
"""Internal matcher — Pick A core algorithm.

Reference: docs/architecture/06-OFFCHAIN-SERVICES-SPEC.md §2.3
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from degenbot.decision.types import MatchPair
from degenbot.matching.price_compat import (
    clearing_price,
    fill_amount,
    is_price_compatible,
)

if TYPE_CHECKING:
    from collections.abc import Generator, Sequence

    from degenbot.decision.types import MatchCandidate


def find_matches(
    outbound: Sequence[MatchCandidate],
    inbound: Sequence[MatchCandidate],
    uniswapx: Sequence[MatchCandidate],
) -> Generator[MatchPair, None, None]:
    """Yields all viable internal-match pairs from the current queue snapshot."""
    counters = list(inbound) + list(uniswapx)

    for o in outbound:
        for c in counters:
            if not is_price_compatible(o, c):
                continue

            fill = fill_amount(o, c)
            if fill == 0:
                continue

            yield MatchPair(
                o=o,
                c=c,
                fill_amount=fill,
                clearing_price=clearing_price(o, c),
            )


def find_best_match(
    outbound: Sequence[MatchCandidate],
    inbound: Sequence[MatchCandidate],
    uniswapx: Sequence[MatchCandidate],
) -> MatchPair | None:
    """Find the single best match by greedy fill_amount maximization."""
    best: MatchPair | None = None
    for m in find_matches(outbound, inbound, uniswapx):
        if best is None or m.fill_amount > best.fill_amount:
            best = m
    return best

File: rust/src/matching/internal_matcher.rs
use crate::matching::price_compat::{clearing_price, fill_amount, is_price_compatible};
use crate::monitor::{MatchCandidate, MatchPair};

pub fn find_matches(outbound: &[MatchCandidate], counters: &[MatchCandidate]) -> Vec<MatchPair> {
    let mut matches = Vec::new();
    for o in outbound {
        for c in counters {
            if !is_price_compatible(o, c) {
                continue;
            }

            let fill = fill_amount(o, c);
            if fill.is_zero() {
                continue;
            }

            matches.push(MatchPair {
                o: o.clone(),
                c: c.clone(),
                fill_amount: fill,
                clearing_price: clearing_price(o, c),
            });
        }
    }
    matches
}

pub fn find_best_match(
    outbound: &[MatchCandidate],
    counters: &[MatchCandidate],
) -> Option<MatchPair> {
    let mut best: Option<MatchPair> = None;
    for m in find_matches(outbound, counters) {
        if let Some(ref b) = best {
            if m.fill_amount > b.fill_amount {
                best = Some(m);
            }
        } else {
            best = Some(m);
        }
    }
    best
}
